pkg/ring/lifecycler.go

```diff
@@ -588,6 +589,13 @@ func (i *Lifecycler) initRing(ctx context.Context) error {
 		i.setTokens(tokens)
 
 		level.Info(log.Logger).Log("msg", "existing entry found in ring", "state", i.GetState(), "tokens", len(tokens), "ring", i.RingName)
+
+		// If we flip the instance from LEAVING to ACTIVE and the heartbeat is disabled, we need to update KV here
+		if i.cfg.HeartbeatPeriod == 0 && ringDesc.Ingesters[i.ID].State == LEAVING && instanceDesc.State == ACTIVE {
```

Comment:
Question: are we going to hit the issue #4365 in the case the ingester crashes while in the `PENDING` or `JOINING` state in the ring and then restarts? Could you add unit tests to check that too, please?

Reply:
Sure..

But i think we are fine there as..

```
		if instanceDesc.State == JOINING {
			level.Warn(log.Logger).Log("msg", "instance found in ring as JOINING, setting to PENDING",
				"ring", i.RingName)
			instanceDesc.State = PENDING
			return ringDesc, true, nil
		}
```